images/inference-services/kfserving/run.sh

```diff
@@ -37,25 +40,63 @@ case $PREDICTOR in
     # named '1', create it and copy the tensorflow model to it.
     tensorflow)
         if [ "$(mc find ${model_bucket} --name 1)" = "" ]; then
-            mc cp -r ${model_bucket}/tfmodel/ ${model_bucket}/1
+            mc cp -r "${model_bucket}"/tfmodel/ "${model_bucket}"/1
         fi
-        export RUNTIME_VERSION=$(mc cat ${model_bucket}/requirements.txt | awk -F '=' '/tensorflow/ {print $3}')
+        export RUNTIME_VERSION=$(mc cat "${model_bucket}"/requirements.txt | awk -F '=' '/tensorflow/ {print $3}')
         prediction_url_path="${isvc}:predict"
         ;;
     # kfserving expects the sklearn model file as model.joblib however mlflow
     # saves the model as model.pkl, so if there is no model.joblib, create a
     # copy named model.joblib from model.pkl
     sklearn)
-        if ! mc ls ${model_bucket} | grep -q "model.joblib"; then
-            mc cp ${model_bucket}/model.pkl ${model_bucket}/model.joblib
+        if ! mc ls "${model_bucket}" | grep -q "model.joblib"; then
+            mc cp "${model_bucket}"/model.pkl "${model_bucket}"/model.joblib
         fi
         export PROTOCOL_VERSION="v2"
         prediction_url_path="${isvc}/infer"
         ;;
+    triton)
+        # triton supports multiple serving backends, each has its own directory
+        # structure (see: https://github.com/triton-inference-server/server/blob/main/docs/model_repository.md).
+        export PROTOCOL_VERSION="v2"
+        export RUNTIME_VERSION="21.08-py3"
+        export ARGS="[--strict-model-config=false]"
+        export FUSEML_MODEL="${FUSEML_MODEL}/triton"
+        prediction_url_path="${isvc}/infer"
+        if [ "$(mc find ${model_bucket} --name triton)" = "" ]; then
+            mlmodel="$(mc cat ${model_bucket}/MLmodel)"
+            if echo "${mlmodel}" | grep -q saved_model; then
+                mc cp -r "${model_bucket}"/tfmodel/ "${model_bucket}"/triton/"${isvc}"/1/model.savedmodel
+            elif echo "${mlmodel}" | grep -q onnx; then
+                model_file="$(echo ${mlmodel} | awk '/data:/ { print $2; exit }')"
+                mc cp -r "${model_bucket}"/"${model_file}" "${model_bucket}"/triton/"${isvc}"/1/
+            else
+                echo "ERROR: Only tensorflow 'saved model' nad ONNX format is supported"
+                exit 1
+            fi
+        fi
 esac
 
+RESOURCES_LIMITS="{cpu: 1000m, memory: 2Gi}"
+# set inference service container resources if specified
+if [ -n "${FUSEML_RESOURCES_LIMITS+x}" ]; then
+    RESOURCES_LIMITS="${FUSEML_RESOURCES_LIMITS}"
+fi
+export RESOURCES_LIMITS
+
+new_ifs=true
+if kubectl get inferenceservice/${isvc} > /dev/null 2>&1; then
+    new_ifs=false
+fi
+
 envsubst < /root/template.sh | kubectl apply -f -
 
+# if the inference service already exists, wait for its status to be updated
+# with the new deployment (eventually it will transition to not Ready as it is
+# waiting for the new deployment to be ready)
+if [ "${new_ifs}" = false ] ; then
```

Comment:
this could get ugly on faster systems where you might miss the "not ready" window and instead enter an infinite loop here

Reply:
Good point, however it won't enter an infinite loop.

I have changed the timeout of the wait command from 600 to 30 seconds, so the worst can happen is just a delay of 30 seconds before it proceed.

Reply:
good enough